timeoutMiddleware.go

```diff
@@ -0,0 +1,40 @@
+package gin
+
+import (
+	"context"
+	"net/http"
+	"time"
+
+	contractshttp "github.com/goravel/framework/contracts/http"
+	"github.com/goravel/framework/contracts/config"
+)
+
+// TimeoutMiddleware creates middleware to set a timeout for a request
+func TimeoutMiddleware() contractshttp.Middleware {
+	return func(ctx contractshttp.Context) {
+		timeout := config.GetInt("http.http.drivers.gin.timeout", 3) * time.Second
+		timeoutCtx, cancel := context.WithTimeout(ctx.Context(), timeout)
+		defer cancel()
+
+		// Refreshing the request context with a timeout
+		ctx.WithContext(timeoutCtx)
+
+		// We start executing the request in a new goroutine
+		done := make(chan struct{})
+		go func() {
+			defer close(done)
+			ctx.Request().Next()
+		}()
+
+		select {
+		case <-done:
+			// The request completed before the timeout expired
+		case <-timeoutCtx.Done():
+			// After the timeout expires, return the status 504 Gateway Timeout
+			if timeoutCtx.Err() == context.DeadlineExceeded {
+				ctx.Response().Writer().WriteHeader(http.StatusGatewayTimeout)
+				_, _ = ctx.Response().Writer().Write([]byte("Request timed out"))
+			}
+		}
```

Comment:
I doubt this's unnecessary, when we set `timeCtx` to `gin`, `gin` will handle this automatically. We can test it.

Reply:
I'll trust your commit, where can I read about what you say?

Reply:
I tested on localhost, gin does not cancel the request when WithTimeout expires. for example:
package main

import (
	"context"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
)

func main() {
	router := gin.Default()

	router.Use(func(c *gin.Context) {
		timeoutCtx, cancel := context.WithTimeout(c.Request.Context(), 3*time.Second)
		defer cancel()
		c.Request = c.Request.WithContext(timeoutCtx)
		c.Next()
	})

	router.GET("/long_request", func(c *gin.Context) {
		time.Sleep(6 * time.Second)
		c.String(http.StatusOK, "This should time out")
	})

	router.Run(":8080")
}

curl http://localhost:8080/long_request

the server continues processing the request after the context is canceled

Reply:
Good catch 👍